src/app/models/field/emailField.ts

```diff
@@ -39,6 +39,41 @@ const createEmailFieldSchema = () => {
       type: Boolean,
       default: false,
     },
+    hasAllowedEmailDomains: {
+      type: Boolean,
+      default: false,
+    },
+    allowedEmailDomains: {
+      type: [String],
+      required: false,
+      // If there allowedEmailDomains is empty, then all email domains should be allowed.
+      default: [],
+      validate: {
+        validator: (emailDomains) => {
+          return (
+            !emailDomains.length ||
+            (new Set(emailDomains).size === emailDomains.length &&
+              emailDomains.filter((s) => s.match(/@.+\..+/)).length ===
+                emailDomains.length)
+          )
+        },
+        message: ({ value }) => {
+          const uniqueEmailDomains = new Set(value)
```

Comment:
Should add typing, and can also be shortened with some ES6 magic:

```ts
import { isEmpty } from 'lodash'

...

message: ({ value: emailDomains }: { value: string[] }) => {
  const duplicateDomains = new Set(
    emailDomains.filter((v, i, a) => a.indexOf(v) !== i),
  )
  return isEmpty(duplicateDomains)
      ? 'There are one or more invalid email domains.'
      : 'There are one or more duplicate email domains.'
}
```

Reply:
If we do the `match` validation as described, then this doesn't even need to check for invalid domains; directly return a message

```ts
message: 'There are one or more duplicate email domains.'
```